core/src/main/java/hudson/model/Queue.java

```diff
@@ -2219,10 +2219,20 @@ public String toString() {
         public Calendar timestamp;
 
         public WaitingItem(Calendar timestamp, Task project, List<Action> actions) {
-            super(project, actions, COUNTER.incrementAndGet(), new FutureImpl(project));
+            super(project, actions, incCurrentCounterValue(), new FutureImpl(project));
             this.timestamp = timestamp;
         }
 
+        private static long incCurrentCounterValue() {
+            try {
+                return COUNTER.incrementAndGet();
+            } finally {
+                // Persist the queue state in case Jenkins
+                // dies unexpectedly. JENKINS-30909
+                getInstance().save();
```

Comment:
This method may fail with runtime exception if Jenkins is shutting down. Is it a potential case?